### DISCUSSION TITLE
How to create pages in production url for dynamic pages using getStaticProps

### DISCUSSION THREAD

**Discussion by llibdude** [tinacms/tinacms#2934] [Help] [answered]
Sorry for title, not sure what words to even use to sum up this issue.  Here is the scenario:

1. I have a website, `example.com`
2. The `main` branch is protected, you cannot write directly to it so you much branch, create content, merge branch into main.
3. I am creating event pages in `example.com/event/[slug]`
4. As a Tina user I go to `exampe.com/admin`, use the branch switcher to create a new branch `eventpageexample`, and create a new event page.
5. I save the page, but it's not ready to merge yet, I want to edit it using contextual editing to see the form on the left and the site on the right (my optimal creation/editing experience).  So I navigate to `example.com/event/eventpageexample`.
6. The page 404s forever.  This is because we have navigated to the "real" page on the website and the website is on `main` but content is on `eventpageexample` branch.

Why this is a problem:
1. I want to use `getStaticProps` because I want my existing event pages to be built at runtime. I could of course not use `getStaticProps` but it's not ideal for me.
2. I'm on the `main` branch trying to access content from `eventpageexample` branch, I should be able to fetch the data by calling to `https://content.tinajs.io/content/${process.env.NEXT_PUBLIC_TINA_CLIENT_ID}/github/eventpageexample`.  Unfortunately since branch names can be dynamic I won't know them so I obviously can't hard code them. _I need a way to know the branch at runtime_

Basically I'm looking for help on a solution strategy.  I realize maybe this is more of a React/Nextjs question than Tina potentially.

Here is the relevant code, my `getStaticProps` in `pages/events/[...event].tsx`. I have my `getStaticPaths` with `fallback: blocking`.

Basically because data fetching all happens server side I assume I'm out of luck on what i want to do.  Possible solutions I considered:
1. Set the branch name in URL. Unfortunately `getStaticProps` doesnt respect url params even in the case of fallback
2. Is there someway to know the branch name from the branch switcher? I assume not because this is all server side.
3. Move away from getStaticProps, but then I don't get prebuilt pages and i lose performance, possible seo, etc.

I realize this is a hard question probably with no perfect example but figure I'll ask anyway and see what happens. Thanks!.

```typescript
export const getStaticProps: GetStaticProps = async ({ params }) => {
  const path = `/${(params.event as string[]).join('/')}`

  const relativePath = `${path}.mdx`
  const client = ExperimentalGetTinaClient()
  let tinaProps = await client.getEventsDocument({
    relativePath
  })

  if (!Object.keys(tinaProps.data).length) {
    // use read only tokens to get live data
    const response = await fetch(
      `https://content.tinajs.io/content/${process.env.NEXT_PUBLIC_TINA_CLIENT_ID}/github/${I NEED THE BRANCH NAME}`,
      {
        method: 'POST',
        body: JSON.stringify({ query: tinaProps.query, variables: { relativePath } }),
        headers: {
          'X-API-KEY': process.env.NEXT_PUBLIC_TINA_READONLY_TOKEN,
          'Content-Type': 'application/json'
        }
      }
    )
    tinaProps = await response.json()
    // This page does not exist so 404
    if (!tinaProps.data) {
      return {
        notFound: true
      }
    }
  }

  return {
    props: {
      ...tinaProps,
    }
  }
}
```

**Comment by coreyaus** [comment]
If your site is hosted on Vercel then I think what you need would be `process.env.NEXT_PUBLIC_VERCEL_GIT_COMMIT_REF`, which is one of the [system environment variables](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables) available on that platform. So your code would look something like this:

```javascript
const branch = process.env.NEXT_PUBLIC_VERCEL_GIT_COMMIT_REF;
const tinaClientId = process.env.NEXT_PUBLIC_TINA_CLIENT_ID;
const response = await fetch(
  `https://content.tinajs.io/content/${tinaClientId}/github/${branch}`,
  {
    method: 'POST',
    body: JSON.stringify({ query: tinaProps.query, variables: { relativePath } }),
    headers: {
      'X-API-KEY': process.env.NEXT_PUBLIC_TINA_READONLY_TOKEN,
      'Content-Type': 'application/json'
    }
  }
)
```

There is also info about this in the Tina docs: https://tina.io/docs/tina-cloud/connecting-site/#using-the-deployment-branch

If you're on another platform like Netlify then something similar is possible but requires a custom build plugin 👍

**Reply by llibdude** [reply]
That is actually what I have today, it works great in the case where the preview url in Vercel is on the same branch as the branch I created. 

However in my case I want to use the branch switcher plugin to create a branch, and without changing the domain of my url, start editing or creating a new document using tina.

Even as I wrote that I know it made no sense but it's hard for me to articulate this for some reason.

Basically i want the workflow:
1. login to tina admin
2. create new branch
3. create new document and save
4. navigate to the new document page in edit mode (so i can see the live updates)

I want to do that all in one fluid workflow that does not require changing urls. In order for this to work I would somehow need to know what branch I created while on step #4 from inside the `getStaticProps` method so that I could set the branch name in the content fetching url.

I am fairly confident the answer is "there isn't a good solution" or "don't use static props and rebuild the pages on each request" or "do something hacky like persist the database name somewhere (like a database/flatfile/cookie?) on your own" but curious to see if there are any supported options out there.

I will likely hack together something tomorrow and report back my best effort

**Comment by Perni1984** [comment]
@llibdude I am not 100% sure, but what we did is to write our custom 404  handler for Next.js (https://nextjs.org/docs/advanced-features/custom-error-page#customizing-the-404-page) that renders a 404 page serverside, but on client side checks if the user is logged in into Tina Admin and then fetches the Tina data client side thus creating an editing experience. 

It's hacky, hardwired and comes from a pre Read-Only-Tokens era, but for us "it just works (tm)" right now, and we have the 4 steps workflow that you are describing.

FWIW I'll post our code here:

```typescript
import dynamic from 'next/dynamic';
import { useRouter } from 'next/router';
import { useTina } from 'tinacms/dist/edit-state';
import { ErrorPage404 } from '../components/molecules/404ErrorPage/ErrorPage404';
import { CONSTANTS } from '../constants';
import { isBrowser } from '../shared/helper';
import {
  customGetPagesDocumentDocument,
  customGetSuccessStoriesDocumentDocument,
  getBlogDocumentQuery,
} from '../shared/queries';

export const ClientSide404Error = () => {
  const queries = {
    '/': customGetPagesDocumentDocument,
    '/blog/': getBlogDocumentQuery,
    '/successStories/': customGetSuccessStoriesDocumentDocument,
  };

  const router = useRouter();
  const originalPathBefore404Error = router.asPath;

  const indexForSlicing = originalPathBefore404Error.lastIndexOf('/') + 1;
  const partialPath = originalPathBefore404Error.slice(0, indexForSlicing);

  let document = originalPathBefore404Error.slice(indexForSlicing);

  if (document === '') {
    document = CONSTANTS.indexPage;
  }
  const variables = {
    relativePath: `${document}.md`,
  };

  const { data } = useTina({
    query: queries?.[partialPath],
    variables,
    data: null,
  });

  if (!data) {
    return <ErrorPage404 />;
  }

  switch (partialPath) {
    case '/': {
      const ContentPage = dynamic(() =>
        import('../components/pages/ContentPage').then(
          (mod) => mod.ContentPage,
        ),
      );

      // @ts-ignore - right now we don't have a way yet to generate types for custom queries
      const props = data?.getPagesDocument?.data;

      return <ContentPage props={props} canonicalUrl={''} filename={''} />;
    }

    case '/blog/': {
      const BlogPage = dynamic(() =>
        import('../components/pages/BlogPage').then((mod) => mod.BlogPage),
      );

      // @ts-ignore - right now we don't have a way yet to generate types for custom queries
      const props = data?.blog?.data;

      return <BlogPage props={props} canonicalUrl={''} cardReferences={[]} />;
    }

    default: {
      return <ErrorPage404 />;
    }
  }
};

export default function Page(props) {
  if (isBrowser()) {
    return <ClientSide404Error />;
  }
  return <ErrorPage404 />;
}

**Reply by llibdude** [reply]
Thanks, This looks promising! I'll give it a shot

Edit - Okay I don't think I see how you handle the different branch name, but that's fine this at least gives me an option on how to handle the data fetching if I can figure out how to get the branch name (probalby using https://github.com/tinacms/tinacms/tree/main/packages/%40tinacms/toolkit/src/plugins/branch-switcher) client side.

**Comment by llibdude** [comment, marked as answer]
I have figured out a solution for nextjs,  it's still a little messy but I think it's the best solution possible.

### The key is to leverage [middleware](https://nextjs.org/docs/advanced-features/middleware):
```typescript
import { NextRequest, NextResponse } from 'next/server'

export const middleware = async (req: NextRequest) => {
  // TinaCMS: If a branch name is provided as a url param we want to use that branch to fetch data for editing a page on a different branch than we are on.
  // We need to do this because there are no ways to get url params from getStaticProps currently
  if (req.nextUrl.pathname.startsWith('/events') && req.nextUrl.searchParams.get('branch') !== null) {
    return NextResponse.rewrite(
      `${req.nextUrl.origin}${req.nextUrl.pathname.replace(/\/$/, '')}-tinacmsbranch-${req.nextUrl.searchParams.get(
        'branch'
      )}`
    )
  }
}
```
Essentially we need to let `getStaticProps` know what branch we want to be on. Outside of some complicated solution with persisting the branch in a database somewhere, I have elected to use a url param.  However you aren't allowed to access url params in `getStaticProps` since it only runs on the server.  But, we can use middleware to take a url param and actually convert it to be part of the url path itself.

### My RouteMapping in `schema.tsx`
```typescript
const RouteMapping = new RouteMappingPlugin((collection, document) => {
      const branch = localStorage.getItem('tinacms-current-branch').replaceAll('"', '')
      return `/${collection.name}/${document.sys.filename}?branch=${branch}`
    })
```
Part of this solution is to leverage the branch name TinaCMS stores in localStorage.  *Note* - This can't be relied on, it could change in the future.  When a user navigates to a page from the admin we will add the branch name to a url param, which will get converted to a url path by middleware as described above.

### Lastly use fetch the branch data in getStaticProps
```
export const getStaticProps: GetStaticProps = async ({ params }) => {
  let slug = params.event[params.event.length - 1]

  // PARSE OUT THE BRANCH NAME AND USE IT IF IT EXISTS
  const match = slug.match(/tinacmsbranch-(.*)/)
  let branch = process.env.VERCEL_GIT_COMMIT_REF
  if (match?.[1]) {
    const regex = RegExp(`-${match[0]}`, 'g')
    slug = slug.replace(regex, '')
    branch = match[1]
  }

  const path = `/${[...params.event.slice(0, params.event.length - 1), slug].join('/')}`

  const relativePath = `${path}.mdx`
  const client = ExperimentalGetTinaClient()
  let tinaProps: any;
  if (!branch) {
    tinaProps = await client.getEventsDocument({
      relativePath
    })
  }
  let tinaData: any

  if (!tinaProps || !Object.keys(tinaProps.data).length) {
    // use read only tokens to get live data
    const response = await fetch(
      `https://content.tinajs.io/content/${process.env.NEXT_PUBLIC_TINA_CLIENT_ID}/github/${branch}`,
      {
        method: 'POST',
        // FIXME: Look into using tinaProps.query. But i think if the initial query fails I get nothing
        body: JSON.stringify({ query: tinaProps.query, variables: { relativePath } }),
        headers: {
          'X-API-KEY': process.env.NEXT_PUBLIC_TINA_READONLY_TOKEN,
          'Content-Type': 'application/json'
        }
      }
    )
    tinaData = await response.json()
    // This page does not exist so 404
    if (!tinaData.data) {
      return {
        notFound: true
      }
    }
  }
  
  // Do other stuff
  .....
  ```
  Finally here is the `getStaticProps` part of the page I'm rendering that requires the branch name.  Using a consistent url name I can set up a regex to match to find the branch name if it exists. If there is a branch provided fetch the data usign read only token.
  
  
  I'm sure I could have explained this more clearly but if something is not clear maybe I can answer questions if people have them, or make a small demo app.
